src/InterpreterGenerator.h

```diff
@@ -0,0 +1,650 @@
+ * - https://opensource.org/licenses/UPL
+ * - <souffle root>/licenses/SOUFFLE-UPL.txt
+ */
+
+/************************************************************************
+ *
+ * @file InterpreterGenerator.h
+ *
+ * Declares the Interpreter Generator class. The generator takes an entry
+ * of the RAM program and translate it into an executable InterpreterNode representation
+ * with environment symbol binding in each node.
+ ***********************************************************************/
+
+#include "InterpreterNode.h"
+#include "InterpreterPreamble.h"
+#include "InterpreterRecords.h"
+#include "RamIndexAnalysis.h"
+#include "RamVisitor.h"
+
+#pragma once
+
+namespace souffle {
+
+/*
+ * @class NodeGenerator
+ * @brief Generate an executable InterpreterNode tree based on the RAM tree.
+ *        Each node contains run time information which is necessary for InterpreterEngine to interpreter.
+ */
+class NodeGenerator : public RamVisitor<std::unique_ptr<InterpreterNode>> {
+    using NodePtr = std::unique_ptr<InterpreterNode>;
+    using NodePtrVec = std::vector<NodePtr>;
+
+public:
+    NodeGenerator(RamIndexAnalysis* isa) : isa(isa) {}
+
+    /**
+     * @brief Generate the tree based on given entry.
+     * Return a NodePtr to the root.
+     */
+    NodePtr generateTree(const RamNode& root) {
+        // Encode all relation, indexPos and viewId.
+        visitDepthFirst(root, [&](const RamNode& node) {
+            if (dynamic_cast<const RamQuery*>(&node)) {
+                newQueryBlock();
+            }
+            if (const auto* create = dynamic_cast<const RamCreate*>(&node)) {
+                encodeRelation(create->getRelation());
+            }
+            if (const auto* indexSearch = dynamic_cast<const RamIndexOperation*>(&node)) {
+                encodeIndexPos(*indexSearch);
+                encodeView(indexSearch);
+            } else if (const auto* exists = dynamic_cast<const RamExistenceCheck*>(&node)) {
+                encodeIndexPos(*exists);
+                encodeView(exists);
+            } else if (const auto* provExists = dynamic_cast<const RamProvenanceExistenceCheck*>(&node)) {
+                encodeIndexPos(*provExists);
+                encodeView(provExists);
+            }
+        });
+        // Parse program
+        return visit(root);
+    }
+
+    NodePtr visitNumber(const RamNumber& num) override {
+        return std::make_unique<InterpreterNode>(I_Number, &num);
+    }
+
+    NodePtr visitTupleElement(const RamTupleElement& access) override {
+        return std::make_unique<InterpreterNode>(I_TupleElement, &access);
+    }
+
+    NodePtr visitAutoIncrement(const RamAutoIncrement& inc) override {
+        return std::make_unique<InterpreterNode>(I_AutoIncrement, &inc);
+    }
+
+    NodePtr visitIntrinsicOperator(const RamIntrinsicOperator& op) override {
+        NodePtrVec children;
+        for (const auto& arg : op.getArguments()) {
+            children.push_back(std::move(visit(arg)));
```

Comment:
The move isn't required here, and may be slower.
```suggestion
            children.push_back(visit(arg));
```